From c447fe51408268fcbaf1f8b6c10956f8ab567c2a Mon Sep 17 00:00:00 2001
From: Xabier Rodriguez Calvar <xrcalvar@igalia.com>
Date: Thu, 18 Mar 2010 15:22:22 +0100
Subject: [PATCH 1/2] Set default animate FALSE

---
 gtkhtml/gtkhtml.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/gtkhtml/gtkhtml.c b/gtkhtml/gtkhtml.c
index 6f020c5..21cab24 100644
--- a/gtkhtml/gtkhtml.c
+++ b/gtkhtml/gtkhtml.c
@@ -3586,6 +3586,7 @@ gtk_html_init (GtkHTML* html)
 					 client_notify_monospace_font, html, NULL, &gconf_error);
 
 	gtk_html_construct (html);
+	gtk_html_set_animate (html, FALSE);
 }
 
 GType
-- 
1.7.0

